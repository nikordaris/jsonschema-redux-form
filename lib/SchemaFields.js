'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reduxForm = require('redux-form');

var _lodash = require('lodash');

var _evaluateStyle4 = require('evaluate-style');

var _evaluateStyle5 = _interopRequireDefault(_evaluateStyle4);

var _selectors = require('./selectors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LABEL_PROP = 'title';

function _compare() {
  var obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _selectors.DEFAULT_PREFIX;

  return function (a, b) {
    if ((0, _selectors.hasOrdinal)(obj[a], prefix)) {
      if ((0, _selectors.hasOrdinal)(obj[b], prefix)) {
        var aOrdinal = (0, _selectors.getOrdinal)(obj[a], prefix);
        var bOrdinal = (0, _selectors.getOrdinal)(obj[b], prefix);
        return aOrdinal - bOrdinal;
      }
      return 1;
    }
    if ((0, _selectors.hasOrdinal)(obj[b], prefix)) {
      return -1;
    }

    var vA = (0, _lodash.get)(obj[a], LABEL_PROP, a);
    var vB = (0, _lodash.get)(obj[b], LABEL_PROP, b);
    return +(vA > vB) || +(vA === vB) - 1;
  };
}

var SchemaFields = function (_Component) {
  _inherits(SchemaFields, _Component);

  function SchemaFields() {
    _classCallCheck(this, SchemaFields);

    return _possibleConstructorReturn(this, (SchemaFields.__proto__ || Object.getPrototypeOf(SchemaFields)).apply(this, arguments));
  }

  _createClass(SchemaFields, [{
    key: 'renderFields',
    value: function renderFields(schema, id, parentName) {
      var _this2 = this;

      var _props = this.props,
          styles = _props.styles,
          FormFieldsTag = _props.formFieldsTag,
          prefix = _props.prefix;

      var _evaluateStyle = (0, _evaluateStyle5.default)(styles, schema),
          formFieldsStyle = _evaluateStyle.formFields;

      var properties = schema.properties || {};
      return _react2.default.createElement(
        FormFieldsTag,
        { key: id, id: id, style: formFieldsStyle },
        Object.keys(properties).filter(function (prop) {
          return (0, _selectors.isEditable)(properties[prop], prefix, false);
        }).sort(_compare(schema.properties, prefix)).map(function (prop, idx) {
          return _this2.renderField(properties[prop], idx, prop, schema.required ? schema.required.includes(prop) : false, parentName);
        })
      );
    }
  }, {
    key: 'renderField',
    value: function renderField(fieldSchema, idx, name, required, namespace) {
      var _props2 = this.props,
          styles = _props2.styles,
          widgets = _props2.widgets,
          widgetProps = _props2.widgetProps,
          prefix = _props2.prefix;

      var _evaluateStyle2 = (0, _evaluateStyle5.default)(styles, fieldSchema),
          formFieldStyles = _evaluateStyle2.formField;

      var fieldName = namespace ? namespace + '.' + name : name;
      var label = (0, _lodash.get)(fieldSchema, LABEL_PROP, name);

      if (fieldSchema.type && fieldSchema.type === 'object') {
        return this.renderFields(fieldSchema, label, fieldName);
      }

      var widget = (0, _selectors.getWidget)(fieldSchema, prefix);

      if ((0, _selectors.hasWidget)(fieldSchema, prefix) && (0, _lodash.has)(widgets, widget)) {
        var Widget = (0, _lodash.get)(widgets, widget);
        var widgetProp = (0, _evaluateStyle5.default)((0, _lodash.get)(widgetProps, widget, {}), fieldSchema);

        if ((0, _lodash.isString)(Widget)) {
          return _react2.default.createElement(Widget, _extends({
            styles: formFieldStyles,
            key: idx,
            name: name,
            schema: fieldSchema
          }, widgetProp));
        }

        return _react2.default.cloneElement(Widget, _extends({
          key: idx,
          styles: formFieldStyles,
          schema: fieldSchema
        }, widgetProp));
      }

      return _react2.default.createElement(_reduxForm.Field, { key: idx, label: label, name: fieldName, required: required });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props3 = this.props,
          styles = _props3.styles,
          schema = _props3.schema,
          FormFieldsTag = _props3.formFieldsTag;

      var _evaluateStyle3 = (0, _evaluateStyle5.default)(styles, schema),
          formFieldsStyle = _evaluateStyle3.formFields;

      return _react2.default.createElement(
        FormFieldsTag,
        { style: formFieldsStyle },
        this.renderFields(schema, 'schemaForm')
      );
    }
  }]);

  return SchemaFields;
}(_react.Component);

SchemaFields.defaultProps = {
  styles: {},
  prefix: _selectors.DEFAULT_PREFIX,
  formFieldsTag: 'div'
};
exports.default = SchemaFields;