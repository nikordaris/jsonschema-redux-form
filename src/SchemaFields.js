// @flow
import React, { Component } from 'react';
import { Field } from 'redux-form';
import { has, get, map, isString, isEmpty } from 'lodash';
import evaluateStyle from 'evaluate-style';

import {
  DEFAULT_PREFIX,
  getLabel,
  getEditable,
  getOrdinal,
  hasOrdinal,
  getWidget,
  hasWidget
} from './selectors';

function _compare(obj = {}, prefix = DEFAULT_PREFIX) {
  return (a, b) => {
    if (hasOrdinal(obj[a], prefix)) {
      if (hasOrdinal(obj[b], prefix)) {
        const aOrdinal = getOrdinal(obj[a], prefix);
        const bOrdinal = getOrdinal(obj[b], prefix);
        return aOrdinal - bOrdinal;
      }
      return 1;
    }
    if (hasOrdinal(obj[b], prefix)) {
      return -1;
    }

    const vA = getLabel(obj[a], a);
    const vB = getLabel(obj[b], b);
    return +(vA > vB) || +(vA === vB) - 1;
  };
}

export type SchemaFormStylesType = {
  formFields: (
    schema: SchemaType
  ) => { [string]: string | number } | { [string]: string | number },
  formField: (
    schema: SchemaType
  ) => { [string]: string | number } | { [string]: string | number }
};

export default class SchemaForm extends Component {
  static defaultProps = {
    styles: {},
    prefix: DEFAULT_PREFIX,
    formFieldsTag: 'div'
  };

  props: {
    schema: SchemaType,
    prefix: string,
    styles: SchemaFormStylesType,
    widgets: { [string]: React.Element<*> | string },
    widgetProps: { [string]: { styles: { [string]: any } } },
    formFieldsTag: string
  };

  isSchemaEmpty(schema: SchemaType) {
    const { prefix } = this.props;
    return (
      has(schema, 'properties') &&
      isEmpty(
        map(schema.properties).filter(prop => getEditable(prop, prefix, false))
      )
    );
  }

  renderFields(schema: SchemaType, id: string, parentName?: string) {
    const { styles, formFieldsTag: FormFieldsTag, prefix } = this.props;
    const { formFields: formFieldsStyle } = evaluateStyle(styles, schema);
    const properties = schema.properties || {};
    return (
      <FormFieldsTag key={id} id={id} style={formFieldsStyle}>
        {Object.keys(properties)
          .filter(prop => getEditable(properties[prop], prefix, false))
          .sort(_compare(schema.properties, prefix))
          .map((prop, idx) =>
            this.renderField(
              properties[prop],
              idx,
              prop,
              schema.required ? schema.required.includes(prop) : false,
              parentName
            )
          )}
      </FormFieldsTag>
    );
  }

  renderField(
    fieldSchema: SchemaType,
    idx: number | string,
    name: string,
    required: boolean,
    namespace?: string
  ) {
    const { styles, widgets, widgetProps, prefix } = this.props;
    const { formField: formFieldStyles } = evaluateStyle(styles, fieldSchema);
    const fieldName = namespace ? `${namespace}.${name}` : name;
    const label = getLabel(fieldSchema, prefix, name);

    if (fieldSchema.type && fieldSchema.type === 'object') {
      return this.renderFields(fieldSchema, label, fieldName);
    }

    const widget = getWidget(fieldSchema, prefix);

    if (hasWidget(fieldSchema, prefix) && has(widgets, widget)) {
      const Widget = get(widgets, widget);
      const widgetProp = evaluateStyle(
        get(widgetProps, widget, {}),
        fieldSchema
      );

      if (isString(Widget)) {
        return (
          <Widget
            styles={formFieldStyles}
            key={idx}
            name={name}
            schema={fieldSchema}
            {...widgetProp}
          />
        );
      }

      return React.cloneElement(Widget, {
        key: idx,
        styles: formFieldStyles,
        schema: fieldSchema,
        ...widgetProp
      });
    }

    return (
      <Field key={idx} label={label} name={fieldName} required={required} />
    );
  }

  render() {
    const { styles, schema, formFieldsTag: FormFieldsTag } = this.props;
    const { formFields: formFieldsStyle } = evaluateStyle(styles, schema);

    return (
      <FormFieldsTag style={formFieldsStyle}>
        {this.renderFields(schema, 'schemaForm')}
      </FormFieldsTag>
    );
  }
}
