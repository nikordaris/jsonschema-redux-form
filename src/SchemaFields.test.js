import React, { Component } from 'react';
import renderer from 'react-test-renderer';

import SchemaFields from './index.js';

const schema = {
  title: 'Test',
  type: 'object',
  properties: {
    foo: {
      title: 'Foo',
      type: 'number',
      meta: {
        form: {
          ordinal: 0,
          editable: true,
          widget: 'NumberInputField'
        }
      }
    },
    bar: {
      title: 'Bar',
      type: 'string',
      meta: {
        form: {
          ordinal: 1,
          editable: true,
          widget: 'InputField'
        }
      }
    }
  }
};

const widgets = {
  NumberInputField: <input type="number" />,
  InputField: 'input'
};

describe('Render Schema', () => {
  it('should render', () => {
    const tree = renderer
      .create(<SchemaFields schema={schema} widgets={widgets} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
