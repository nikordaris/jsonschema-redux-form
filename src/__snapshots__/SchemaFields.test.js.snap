// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Schema should render 1`] = `
<div
  style={undefined}
>
  <div
    id="schemaForm"
    style={undefined}
  >
    <input
      schema={
        Object {
          "meta": Object {
            "form": Object {
              "editable": true,
              "ordinal": 0,
              "widget": "NumberInputField",
            },
          },
          "title": "Foo",
          "type": "number",
        }
      }
      styles={undefined}
      type="number"
    />
    <input
      name="bar"
      schema={
        Object {
          "meta": Object {
            "form": Object {
              "editable": true,
              "ordinal": 1,
              "widget": "InputField",
            },
          },
          "title": "Bar",
          "type": "string",
        }
      }
      styles={undefined}
    />
  </div>
</div>
`;
